spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/provider/SpringCacheCacheTest.java

```diff
@@ -212,6 +209,51 @@ public void testValueLoaderWithNoPreviousValue() {
       assertEquals("test", valueFromCache.get());
    }
 
+   /*
+    * In this test Thread 1 should exclusively block Cache#get method so that Thread 2 won't be able to
+    * insert "thread2" string into the cache.
+    *
+    * The test check this part of the Spring spec:
+    * Return the value to which this cache maps the specified key, obtaining that value from valueLoader if necessary.
+    * This method provides a simple substitute for the conventional "if cached, return; otherwise create, cache and return" pattern.
+    * @see http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/cache/Cache.html#get-java.lang.Object-java.util.concurrent.Callable-
+    */
+   @Test(timeOut = 30_000)
+   public void testValueLoaderWithLocking() throws Exception {
+      //given
+      CountDownLatch waitUntilThread1LocksValueGetter = new CountDownLatch(1);
+      CountDownLatch waitUntilBothThreadsFinish = new CountDownLatch(2);
+
+      //when
+      fork(() -> {
+         Thread.currentThread().setName("Thread1");
+         cache.get("test", () -> {
+            waitUntilThread1LocksValueGetter.countDown();
+            TimeUnit.MILLISECONDS.sleep(10);
+            return "thread1";
+         });
+         waitUntilBothThreadsFinish.countDown();
+      });
+
+      fork(() -> {
+         Thread.currentThread().setName("Thread2");
+         try {
+            waitUntilThread1LocksValueGetter.await();
+         } catch (InterruptedException e) {
+            throw new AssertionError(e);
+         }
+         cache.get("test", () -> "thread2");
```

Comment:
@slaskawi The synchronization might be correct, but after Thread 2 being unblocked it might not see the value `thread1` and run the supplier. The supplier value could not be written for some reason, but it would be returned.

Or, the synchronization could not work and Thread 1 wouldn't see the updated value either, ending up with value `thread 1`.